src/java.base/share/classes/sun/launcher/LauncherHelper.java

```diff
@@ -596,6 +597,9 @@ static String getMainClassFromJar(String jarname) {
             if (opens != null) {
                 addExportsOrOpens(opens, true);
             }
+            if (Boolean.parseBoolean(mainAttrs.getValue(ENABLE_NATIVE_ACCESS))) {
```

Comment:
From a design perspective there are two options here:
* `false` means "ignore" (which is what is being implemented here)
* `false` means `no native access`.

The latter seems a bit odd - but I'd like to remind the behavior of what happens when `--enable-native-access foo` is specified, and access to restricted method occurs in a module other than `foo`. In that case we throw an exception (e.g. we validate strictly, because a command line option was found).

Jar could do a similar thing:
* if no manifest attribute is specified, do the current warning behavior
* if attribute is set to `true`, do not warn
* if attribute is set to `false`, access to restricted methods in this jar are simply not allowed and must fail with exception

Reply:
I had a chat with @AlanBateman. He agrees that seeing "Enable-Native-Access:false" strongly suggests of a deny-like behavior (e.g. access should fail). That said, if we just did that, then it would be possible to have an `--enable-native-access` command line option and a manifest attribute pulling in different directions (e.g. enable vs. disable).

Ideally, we'd like a value-less attribute (e.g. `EnableNativeAccess:`) so that there's no confusion re. true/false. But that is not allowed by the manifest attribute grammar.

The next best thing seems to be `EnableNativeAccess:ALL-UNNAMED`. E.g. we only allow the name of the module we're allowed to enable native access for (the unnamed module). If, in the future, jars will start supporting multiple modules, we can enhance the attribute to accept a comma separated list.

This is slightly less confusing. There's no "deny" state, and no possibilities for the attribute to conflict with the command line option.

Reply:
> but I'd like to remind the behavior of what happens when --enable-native-access foo is specified
> if attribute is set to false, access to restricted methods in this jar are simply not allowed and must fail with exception

Right... I agree here.

I've gone ahead and implemented this in the latest commit. This is a bit tricky since we initialize ModuleBootstrap, which holds the 'has enable native access' flag before we process the manifest. I've implemented it as a mutable static that gets lazily set in LauncherHelper.

WRT the tests. In order to test this I'm launching an executable jar from the class path, which then calls into a module that tries to do the native access.

But, I'm now starting to wonder whether the `true`/`false` value is really good enough... Does it make sense that a class path jar referencing several modules might want to enable native access for them on a per-module basis? Is an executable jar referencing a module even a case we want to care about at all?

Reply:
> The next best thing seems to be EnableNativeAccess:ALL-UNNAMED. E.g. we only allow the name of the module we're allowed to enable native access for (the unnamed module). If, in the future, jars will start supporting multiple modules, we can enhance the attribute to accept a comma separated list.

> This is slightly less confusing. There's no "deny" state, and no possibilities for the attribute to conflict with the command line option.

Didn't see these comments. This sounds like a good plan. Will work on that next.

Reply:
> do we want to allow module names for modules outside of the executable jar file? i.e. should it just be a supplement to the --enable-native-access command line option?

After thinking about this for a bit, I think we should allow multiple module names. Consider for instance an executable jar which depends on a library, inside another modular jar file. The library jar is put on the module path, the executable jar passed to `-jar`. I think there should be a way in this case in the jar manifest to enable native access for the library if it needs it. If we just allow `EnableNativeAccess: ALL-UNNAMED` than that doesn't seem possible?

Reply:
> > do we want to allow module names for modules outside of the executable jar file? i.e. should it just be a supplement to the --enable-native-access command line option?
> 
> After thinking about this for a bit, I think we should allow multiple module names. Consider for instance an executable jar which depends on a library, inside another modular jar file. The library jar is put on the module path, the executable jar passed to `-jar`. I think there should be a way in this case in the jar manifest to enable native access for the library if it needs it. If we just allow `EnableNativeAccess: ALL-UNNAMED` than that doesn't seem possible?

The intent in my previous comment was to only allow ALL-UNNAMED as possible value. I believe this is consistent with the Add-Opens attribute which also targets the unnamed module:

> A <module>/<package> pair in the value of an Add-Opens attribute has the same meaning as the command-line option --add-opens <module>/<package>=ALL-UNNAMED.

So, there is a strong precedent here for the attribute in the manifest of a jar **not** to affect other modules. If other modules are present, a command line option needs to be added too.

Reply:
> So, there is a strong precedent here for the attribute in the manifest of a jar **not** to affect other modules. If other modules are present, a command line option needs to be added too.

Yes, I think limit it for now to the value "ALL-UNNAMED". Maybe in the future there will be OOTB support for having multiple modules in the same JAR file or executable modular JARs, at which point this might have to looked at again.